# Laboratorio 2 Data Science
- Kenneth Gálvez 20079
- José Mariano Reyes 20074

## 1. Preparación de datos

In [1]:
import numpy as np
import pickle
from sklearn.model_selection import train_test_split
from tensorflow.keras.utils import to_categorical

# Cargar los datos desde los archivos .p
with open('entrenamiento.p', 'rb') as f:
    train_data = pickle.load(f)
    
with open('validacion.p', 'rb') as f:
    validation_data = pickle.load(f)
    
with open('prueba.p', 'rb') as f:
    test_data = pickle.load(f)

# Dividir los datos en imágenes y etiquetas
x_train, y_train = train_data['features'], train_data['labels']
x_val, y_val = validation_data['features'], validation_data['labels']
x_test, y_test = test_data['features'], test_data['labels']

# Preprocesamiento: Redimensionar y Normalizar
input_shape = (32, 32, 3)  # Por ejemplo, redimensionar a 32x32 y 3 canales de color (RGB)
x_train = x_train / 255.0  # Normalización
x_val = x_val / 255.0
x_test = x_test / 255.0

# Preprocesamiento: Codificación One-Hot para las etiquetas
num_classes = 43
y_train = to_categorical(y_train, num_classes)
y_val = to_categorical(y_val, num_classes)
y_test = to_categorical(y_test, num_classes)


## 2. Implementación de la arquitectura Le-Net

### Arquitectura LeNet:

La arquitectura LeNet fue propuesta por Yann LeCun en 1998 y es considerada una de las primeras arquitecturas exitosas de redes neuronales convolucionales. Fue diseñada para el reconocimiento de caracteres escritos a mano y, aunque es relativamente simple en comparación con las CNN modernas, sienta las bases para muchas de las arquitecturas posteriores. La arquitectura LeNet consta de las siguientes capas:

1. Capa de Convolución (C1): Convolución 2D con filtros (kernels) de tamaño pequeño. Aplicación de función de activación, típicamente ReLU (Rectified Linear Unit).Captura características locales en la imagen, como bordes y texturas.

2. Capa de Sub-muestreo (Pooling) (S2): Operación de sub-muestreo, como MaxPooling, para reducir el tamaño de la imagen. Ayuda a conservar las características más importantes y reduce la cantidad de parámetros.

3. Capa de Convolución (C3): Otra capa de convolución con filtros más grandes. Función de activación ReLU. Captura características más abstractas basadas en las características capturadas en C1.

4. Capa de Sub-muestreo (Pooling) (S4): Otra operación de sub-muestreo para reducir el tamaño aún más.

5. Capa completamente conectada (Fully Connected) (F5): Capa densamente conectada con unidades de activación ReLU. Captura relaciones más complejas entre las características capturadas en capas anteriores.

6. Capa completamente conectada (Fully Connected) (F6): Capa final que produce las salidas finales de clasificación. Puede tener unidades de activación lineales o softmax, dependiendo del problema.

### Diagrama de la Red LeNet (png entregado aparte)

### Proceso de Convolución, Función de Activación y Pooling:

1. Convolución: En la capa de convolución, los filtros se deslizan sobre la imagen de entrada realizando productos escalares locales. Esto ayuda a detectar patrones específicos como bordes, texturas y características relevantes en la imagen.

2. Función de Activación: Después de cada operación de convolución, se aplica una función de activación, como la función ReLU. ReLU convierte los valores negativos en cero y mantiene los valores positivos sin cambios. Esto introduce no linealidades en la red, lo que le permite capturar relaciones más complejas.

3. Pooling (Sub-muestreo): La operación de pooling reduce el tamaño espacial de la representación de la imagen mientras mantiene las características más importantes. MaxPooling, por ejemplo, selecciona el valor máximo dentro de una ventana en la imagen. Esto ayuda a reducir la cantidad de parámetros y hacer que la red sea más robusta a pequeñas transformaciones en la imagen.

## 3. Construcción del modelo

In [2]:
import tensorflow as tf
from tensorflow.keras import layers, models

# Definición del modelo LeNet
model = models.Sequential()

# Capa de Convolución (C1)
model.add(layers.Conv2D(filters=6, kernel_size=(5, 5), activation='relu', input_shape=(32, 32, 3)))
# Capa de Sub-muestreo (S2)
model.add(layers.MaxPooling2D(pool_size=(2, 2)))

# Capa de Convolución (C3)
model.add(layers.Conv2D(filters=16, kernel_size=(5, 5), activation='relu'))
# Capa de Sub-muestreo (S4)
model.add(layers.MaxPooling2D(pool_size=(2, 2)))

# Aplanar los datos para las capas Fully Connected
model.add(layers.Flatten())

# Capa Fully Connected (F5)
model.add(layers.Dense(units=120, activation='relu'))

# Capa Fully Connected (F6)
model.add(layers.Dense(units=84, activation='relu'))

# Capa de Salida
model.add(layers.Dense(units=43, activation='softmax'))  # 43 clases en este caso

# Compilar el modelo
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])


## 4. Entrenamiento del modelo

In [3]:
# Definir hiperparámetros
batch_size = 64
epochs = 10

# Entrenar el modelo
history = model.fit(x_train, y_train, batch_size=batch_size, epochs=epochs,
                    validation_data=(x_val, y_val))


Epoch 1/10
544/544 [==============================] - 8s 13ms/step - loss: 1.4440 - accuracy: 0.6232 - val_loss: 0.7442 - val_accuracy: 0.8145
Epoch 2/10
544/544 [==============================] - 6s 11ms/step - loss: 0.3671 - accuracy: 0.9025 - val_loss: 0.5995 - val_accuracy: 0.8388
Epoch 3/10
544/544 [==============================] - 6s 11ms/step - loss: 0.2108 - accuracy: 0.9447 - val_loss: 0.5406 - val_accuracy: 0.8503
Epoch 4/10
544/544 [==============================] - 6s 11ms/step - loss: 0.1448 - accuracy: 0.9633 - val_loss: 0.4810 - val_accuracy: 0.8828
Epoch 5/10
544/544 [==============================] - 6s 11ms/step - loss: 0.1065 - accuracy: 0.9717 - val_loss: 0.3554 - val_accuracy: 0.9132
Epoch 6/10
544/544 [==============================] - 6s 11ms/step - loss: 0.0838 - accuracy: 0.9779 - val_loss: 0.4039 - val_accuracy: 0.9116
Epoch 7/10
544/544 [==============================] - 6s 11ms/step - loss: 0.0689 - accuracy: 0.9823 - val_loss: 0.3894 - val_accuracy: 0.9134

## 5. Evaluación del modelo

In [4]:
from sklearn.metrics import accuracy_score, f1_score

# Predicciones en el conjunto de prueba
y_pred = model.predict(x_test)
y_pred_classes = np.argmax(y_pred, axis=1)

# Etiquetas reales en el conjunto de prueba
y_true = np.argmax(y_test, axis=1)

# Calcular métricas
accuracy = accuracy_score(y_true, y_pred_classes)
f1 = f1_score(y_true, y_pred_classes, average='weighted')

print(f"Precisión en el conjunto de prueba: {accuracy}")
print(f"F1-Score en el conjunto de prueba: {f1}")


395/395 [==============================] - 2s 4ms/step
Precisión en el conjunto de prueba: 0.9079968329374505
F1-Score en el conjunto de prueba: 0.9074452517726638


# Interpretación de Resultados

El modelo muestra un desempeño bueno en los conjuntos de entrenamiento y validación, con una precisión cercana al 98% después de 10 épocas. Pero en el conjunto de prueba, la precisión es fue un poco más baja, alrededor del 90.8%. Creemos que este desajuste se debe a un sobreajuste en los datos de entrenamiento. Igual consideramos que este resultado es bastante impresionante, considerando la complejidad y la variabilidad de las señales de tráfico en las 43 clases.

El F1-Score también fue es alto, lo que significa que el modelo tiene un buen equilibrio entre la precisión y la capacidad para recuperar las instancias relevantes de cada clase. Al analizar el desempeño por clase, observamos que algunas clases tienen una precisión y un F1-Score notablemente altos, mientras que otras pueden tener un rendimiento menor. Esto podría deberse a desafíos específicos de cada clase, como el tamaño de la muestra o la complejidad de las señales. 

## 6. Mejoras y Experimentación

# Experimentos Realizados:
En un esfuerzo por mejorar el rendimiento del modelo, se realizó un experimento utilizando aumento de datos. Se aplicó una técnica de aumento de datos que incluye rotación, desplazamiento y reflejo horizontal a las imágenes del conjunto de entrenamiento. El objetivo era aumentar la variabilidad de los datos y hacer que el modelo sea más resistente a las variaciones en las imágenes.

In [3]:
import numpy as np
import pickle
import tensorflow as tf
from sklearn.model_selection import train_test_split
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from sklearn.metrics import accuracy_score, f1_score
from tensorflow.keras import layers, models

# Cargar los datos desde los archivos .p
with open('entrenamiento.p', 'rb') as f:
    train_data = pickle.load(f)
    
with open('validacion.p', 'rb') as f:
    validation_data = pickle.load(f)
    
with open('prueba.p', 'rb') as f:
    test_data = pickle.load(f)

# Dividir los datos en imágenes y etiquetas
x_train, y_train = train_data['features'], train_data['labels']
x_val, y_val = validation_data['features'], validation_data['labels']
x_test, y_test = test_data['features'], test_data['labels']

# Preprocesamiento: Redimensionar y Normalizar
input_shape = (32, 32, 3)
x_train = x_train / 255.0
x_val = x_val / 255.0
x_test = x_test / 255.0

# Codificación One-Hot para las etiquetas
num_classes = 43
y_train = to_categorical(y_train, num_classes)
y_val = to_categorical(y_val, num_classes)
y_test = to_categorical(y_test, num_classes)

# Experimento: Aumento de Datos
datagen = ImageDataGenerator(
    rotation_range=10,  # Rotación máxima de 10 grados
    width_shift_range=0.1,  # Desplazamiento horizontal máximo del 10% del ancho
    height_shift_range=0.1,  # Desplazamiento vertical máximo del 10% de la altura
    horizontal_flip=True  # Reflejo horizontal aleatorio
)

# Crear un generador de datos para el conjunto de entrenamiento
datagen.fit(x_train)

# Definir hiperparámetros para el entrenamiento con aumento de datos
batch_size = 64
epochs = 20

# Definir el modelo LeNet con Keras

# Definición del modelo LeNet
model = models.Sequential()

# Capa de Convolución (C1)
model.add(layers.Conv2D(filters=6, kernel_size=(5, 5), activation='relu', input_shape=(32, 32, 3)))
# Capa de Sub-muestreo (S2)
model.add(layers.MaxPooling2D(pool_size=(2, 2)))

# Capa de Convolución (C3)
model.add(layers.Conv2D(filters=16, kernel_size=(5, 5), activation='relu'))
# Capa de Sub-muestreo (S4)
model.add(layers.MaxPooling2D(pool_size=(2, 2)))

# Aplanar los datos para las capas Fully Connected
model.add(layers.Flatten())

# Capa Fully Connected (F5)
model.add(layers.Dense(units=120, activation='relu'))

# Capa Fully Connected (F6)
model.add(layers.Dense(units=84, activation='relu'))

# Capa de Salida
model.add(layers.Dense(units=43, activation='softmax'))  # 43 clases en este caso

# Compilar el modelo
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# Compilar el modelo
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# Entrenamiento del modelo con aumento de datos
history = model.fit(datagen.flow(x_train, y_train, batch_size=batch_size),
                    steps_per_epoch=len(x_train) // batch_size,
                    epochs=epochs, validation_data=(x_val, y_val))

# Evaluar el modelo en el conjunto de prueba
y_pred = model.predict(x_test)
y_pred_classes = np.argmax(y_pred, axis=1)
y_true = np.argmax(y_test, axis=1)

accuracy = accuracy_score(y_true, y_pred_classes)
f1 = f1_score(y_true, y_pred_classes, average='weighted')

print(f"Precisión en el conjunto de prueba: {accuracy}")
print(f"F1-Score en el conjunto de prueba: {f1}")


Epoch 1/20
543/543 [==============================] - 13s 23ms/step - loss: 2.4339 - accuracy: 0.3149 - val_loss: 1.6841 - val_accuracy: 0.4558
Epoch 2/20
543/543 [==============================] - 13s 23ms/step - loss: 1.3879 - accuracy: 0.5552 - val_loss: 1.3539 - val_accuracy: 0.5664
Epoch 3/20
543/543 [==============================] - 14s 27ms/step - loss: 1.0485 - accuracy: 0.6575 - val_loss: 1.2008 - val_accuracy: 0.6188
Epoch 4/20
543/543 [==============================] - 14s 26ms/step - loss: 0.8509 - accuracy: 0.7207 - val_loss: 1.0730 - val_accuracy: 0.6553
Epoch 5/20
543/543 [==============================] - 13s 24ms/step - loss: 0.7045 - accuracy: 0.7669 - val_loss: 1.0479 - val_accuracy: 0.6744
Epoch 6/20
543/543 [==============================] - 13s 23ms/step - loss: 0.6177 - accuracy: 0.7968 - val_loss: 0.9479 - val_accuracy: 0.7358
Epoch 7/20
543/543 [==============================] - 12s 23ms/step - loss: 0.5350 - accuracy: 0.8235 - val_loss: 0.9131 - val_accuracy:

# Reflexión sobre Resultados:
Los resultados del experimento que involucró la técnica de aumento de datos muestran un desempeño ligeramente diferente en comparación con el modelo anterior. A pesar de que se obtuvo una precisión en el conjunto de prueba de alrededor del 83.6% y un F1-Score de aproximadamente 83.4%, estos valores son más bajos que los obtenidos en el modelo original sin modificaciones drásticas.

Concluimos que en este caso particular, la aplicación de técnicas de aumento de datos no necesariamente mejoró el rendimiento del modelo de manera positiva, Aunque la idea detrás del aumento de datos es incrementar la variabilidad de los datos de entrenamiento y mejorar la capacidad de generalización del modelo era buena, pues es importante recordar que no todas las técnicas funcionarán igualmente bien para todos los conjuntos de datos.

Es posible que las señales de tráfico tengan características intrínsecas que no se beneficiaron significativamente del aumento de datos en este experimento. Los desafíos específicos de algunas clases pueden requerir enfoques más específicos para mejorar su rendimiento. 

La leccion mas importantes que nos llevamos de este laboratorio es que en situaciones del mundo real, estos resultados nos recuerdan la importancia de explorar diversas estrategias y ajustarlas según las características particulares del conjunto de datos. La adaptación y el ajuste cuidadoso de las técnicas de mejora del rendimiento son esenciales para lograr un modelo de clasificación efectivo y robusto en diferentes situaciones prácticas.